pkg/chunk/aws/dynamodb_storage_client.go

```diff
@@ -587,11 +587,17 @@ type dynamoDBReadResponse struct {
 	items []map[string]*dynamodb.AttributeValue
 }
 
-func (b *dynamoDBReadResponse) Iterator() chunk.ReadBatchIterator {
-	return &dynamoDBReadResponseIterator{
-		i:                    -1,
-		dynamoDBReadResponse: b,
+func (b *dynamoDBReadResponse) Iterator(iter chunk.ReadBatchIterator) chunk.ReadBatchIterator {
+	if iter == nil {
+		return &dynamoDBReadResponseIterator{
+			i:                    -1,
+			dynamoDBReadResponse: b,
+		}
 	}
+	concrete := iter.(*dynamoDBReadResponseIterator)
```

Comment:
This will panic when used with different implementation. Following lines will panic if `concrete` is `nil`, but `iter` was not `nil`.